pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/funnel/FunnelMaxStepAggregationFunction.java

```diff
@@ -0,0 +1,302 @@
+import org.apache.pinot.core.query.aggregation.function.AggregationFunction;
+import org.apache.pinot.core.query.aggregation.groupby.GroupByResultHolder;
+import org.apache.pinot.core.query.aggregation.groupby.ObjectGroupByResultHolder;
+import org.apache.pinot.segment.spi.AggregationFunctionType;
+
+
+public class FunnelMaxStepAggregationFunction
+    implements AggregationFunction<PriorityQueue<FunnelStepEvent>, Long> {
+  private final ExpressionContext _timestampExpression;
+  private final long _windowSize;
+  private final List<ExpressionContext> _stepExpressions;
+  private final Set<Mode> _modes = new HashSet<>();
+  private final int _numSteps;
+
+  public FunnelMaxStepAggregationFunction(List<ExpressionContext> arguments) {
+    int numArguments = arguments.size();
+    Preconditions.checkArgument(numArguments > 2,
+        "FUNNELMAXSTEP expects >= 3 arguments, got: %s. The function can be used as "
+            + "funnelMaxStep(timestampExpression, windowSize, ARRAY[stepExpression, ..], [mode, [mode, ... ]])",
+        numArguments);
+    _timestampExpression = arguments.get(0);
+    _windowSize = arguments.get(1).getLiteral().getLongValue();
+    Preconditions.checkArgument(_windowSize > 0, "Window size must be > 0");
+    ExpressionContext stepExpressionContext = arguments.get(2);
+    if (stepExpressionContext.getFunction() != null) {
+      // LEAF stage init this function like funnelmaxstep($1,'1000',arrayValueConstructor($2,$3,$4,...))
+      _stepExpressions = stepExpressionContext.getFunction().getArguments();
+    } else {
+      // Intermediate Stage init this function like funnelmaxstep($1,'1000',__PLACEHOLDER__)
+      _stepExpressions = ImmutableList.of();
+    }
+    if (numArguments > 3) {
+      arguments.subList(3, numArguments).forEach(arg -> _modes.add(Mode.valueOf(arg.getLiteral().getStringValue())));
+    }
+    _numSteps = _stepExpressions.size();
+  }
+
+  @Override
+  public AggregationFunctionType getType() {
+    return AggregationFunctionType.FUNNELMAXSTEP;
+  }
+
+  @Override
+  public String getResultColumnName() {
+    return getType().getName().toLowerCase() + "(" + _windowSize + ")  (" + _timestampExpression.toString() + ", "
+        + _stepExpressions.stream().map(ExpressionContext::toString)
+        .collect(Collectors.joining(",")) + ")";
+  }
+
+  @Override
+  public List<ExpressionContext> getInputExpressions() {
+    final List<ExpressionContext> inputs = new ArrayList<>();
+    inputs.add(_timestampExpression);
+    inputs.addAll(_stepExpressions);
+    return inputs;
+  }
+
+  @Override
+  public AggregationResultHolder createAggregationResultHolder() {
+    return new ObjectAggregationResultHolder();
+  }
+
+  @Override
+  public GroupByResultHolder createGroupByResultHolder(int initialCapacity, int maxCapacity) {
+    return new ObjectGroupByResultHolder(initialCapacity, maxCapacity);
+  }
+
+  @Override
+  public void aggregate(int length, AggregationResultHolder aggregationResultHolder,
```

Comment:
What is the expected behavior when there are `null` values?

Reply:
The function computed the `FunnelStepEvent` first, so null values are ignored always.